### PR TITLE
ORM: Cache the logger adapter for `ProcessNode`

### PR DESCRIPTION
**Pull request by sphuber**
The logger adapter was recreated each time the `logger` property of the `ProcessNode` was invoked. It is now cached on the node instance as its contents should not change for the lifetime of the node.

### CODE REVIEW

File: src/aiida/orm/nodes/process/process.py [hunk 1]
```diff
@@ -162,6 +162,7 @@ class ProcessNode(Sealable, Node):
     METADATA_INPUTS_KEY: str = 'metadata_inputs'
 
     _unstorable_message = 'only Data, WorkflowNode, CalculationNode or their subclasses can be stored'
+    _logger_adapter = None
```
Comment (khsrali) [inline src/aiida/orm/nodes/process/process.py:165]: is this a class variable or an instance variable?

Comment (sphuber) [reply src/aiida/orm/nodes/process/process.py]: It is a class variable, but you are right it should be an instance variable. I have updated the code.

Comment (sphuber) [reply src/aiida/orm/nodes/process/process.py]: @khsrali  Ok, it turned out it was a bit more difficult. We cannot initialize the `_logger_adapter` in the constructor, because that path is not followded when loading an existing node from the database. So we _have_ initialize it in the `logger` property and initialize it as `None` as a class variable. Since `logger` sets the adapter on the instance, it won't pollute the class variable though, so it seems fine as a solution.

File: src/aiida/orm/nodes/process/process.py [hunk 2]
```diff
@@ -251,7 +259,16 @@ def logger(self):
...
         from aiida.orm.utils.log import create_logger_adapter
 
-        return create_logger_adapter(self._logger, self)
+        # If the node is not yet stored, there is no point in creating the logger adapter yet as the ``DbLogHandler``
+        # it configures only is triggered for stored nodes otherwise it cannot link the log message to the node.
+        if not self.pk:
+            return self._logger
+
+        # First time the property is called after the node is stored, create the logger adapter
+        if self._logger_adapter is None:
```
Comment (khsrali) [inline src/aiida/orm/nodes/process/process.py:268]: Although it works, how about we change this condition to:

```suggestion
        if not hasattr(self, '_logger_adapter'):
```
Then, there is no need to define a class variable.

it seems `if self._logger_adapter is None:` first, checks for instance variables; if they cannot be found, it looks for class variables. In the end, it sets instance variables.

File: src/aiida/orm/nodes/process/process.py [hunk 3]
```diff
@@ -251,7 +253,17 @@ def logger(self):
...
         from aiida.orm.utils.log import create_logger_adapter
 
-        return create_logger_adapter(self._logger, self)
+        # If the node is not yet stored, there is no point in creating the logger adapter yet, as the ``DbLogHandler``
+        # it configures, is only triggered for stored nodes, otherwise it cannot link the log message to the node.
+        if not self.pk:
+            return self._logger
+
+        # First time the property is called after the node is stored, create the logger adapter
+        # if self._logger_adapter is None:
```
Comment (khsrali) [inline src/aiida/orm/nodes/process/process.py:262]:
```suggestion
```

File: src/aiida/orm/nodes/process/process.py [hunk 4]
```diff
@@ -8,6 +8,8 @@
 ###########################################################################
 """Module with `Node` sub class for processes."""
 
+from __future__ import annotations
```
Comment (khsrali) [inline src/aiida/orm/nodes/process/process.py:11]: I'm not sure why is this needed
```suggestion
```

**Review by khsrali — APPROVED** [review]
thanks a lot @sphuber :+1:

File: src/aiida/orm/nodes/process/process.py (hunk 4 above)
Comment (sphuber) [reply src/aiida/orm/nodes/process/process.py:11]: This is a remnant for when I added the type annotations `logging.LoggerAdapater | None`. The `|` for `Union` is only built-in as of Python 3.10. For older versions you have to enable it with this statement.
